### PR TITLE
#160056369 Setup React Router

### PR DESCRIPTION
**Pull request by JCanaks**
#### What does this PR do?
It setups react router version 4 (React-router-dom) for this application
#### Description of Task to be completed?
- Create a new branch from the updated staging branch  and name according to task assigned
- install react router dependencies
- configure all key project routes
- set up views/components folder structure for all route views and components
- create dummy route components
- configure authentication for all protected routes
#### How should this be manually tested?
- Clone the repository to get a local copy
- Run `npm install` to install the necessary dependencies
- Run `npm run dev` to start the app
- Test the following routes 
  -  `/signup`- Home 
  -  `/login` - Login 
  -  `/articles` -All Articles of a user
  - `/articles/:slug`- An article of a user
  - `/articles/:new`- To create a new article (Authentication Required) 
  - `/articles/:slug/edit`- To edit an article(Authentication Required)
  - `/@:username` - To get the profile of a user
  - `/:username/edit` - To edit a user profile (Authentication Required)
  - `/:username/following` - To view the users a person is following
  - `/:username/followers` -  To view the followers of a user
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
#160056369
#### Screenshots (if appropriate)
**Sample screenshot of some routes** 

screen shot 2018-09-03 at 11 48 34 am
screen shot 2018-09-03 at 11 49 02 am
screen shot 2018-09-03 at 11 49 13 am
screen shot 2018-09-03 at 11 51 04 am
screen shot 2018-09-03 at 11 51 28 am
screen shot 2018-09-03 at 11 53 08 am

#### Questions:
N/A

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -1,10 +1,12 @@
 {
     "extends": "airbnb",
+    "parser": "babel-eslint",
     "parserOptions": { 
       "ecmaVersion": 6,
       "sourceType": "module",
       "ecmaFeatures": {
-        "jsx": true
+        "jsx": true,
+        "experimentalObjectRestSpread": true
```
Comment (Fattylee) [inline .eslintrc.json:9]: why is this property added?

**Review by Fattylee — APPROVED** [review]
A job well-done. keep it up. Don't forget to attend to Hound violations.

File: semantic.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+      "compressed": "dist/components/",
+      "themes": "dist/themes/"
+    },
+    "clean": "dist/"
+  },
+  "permission": false,
+  "autoInstall": false,
+  "rtl": false,
+  "version": "2.3.3"
+}
```
Comment (megame24) [inline semantic.json]: EOL

File: src/views/articles/subcategory/SingleArticle.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import React from 'react';
+
+const Article = () => (
+  <div>
+    <h2>An Article in a subcategory</h2>
```
Comment (megame24) [inline src/views/articles/subcategory/SingleArticle.js]: Do we need a separate page to display an article in a subcategory, shouldn't this be the same with a single article's page?

File: src/views/auth/AuthRoute.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Route, Redirect } from 'react-router-dom';
+
+const AuthService = {
```
Comment (megame24) [inline src/views/auth/AuthRoute.js:5]: What are you trying to achieve with this authService?

File: .eslintrc.json (hunk 1 above)
Comment (JCanaks) [reply .eslintrc.json:9]: It helps resolve an Eslint error concerning object spreading

File: src/views/auth/AuthRoute.js (hunk 4 above)
Comment (JCanaks) [reply src/views/auth/AuthRoute.js:5]: It is part of the configuration for checking if a user is authenticated or not. If a user is not logged in and tries to visit a route that is authenticated it redirects him to the login route.  If you view the login component you will see its usage. Once a person successfully logs in they are authenticated and can view all protected routes.

File: src/views/articles/subcategory/SingleArticle.js (hunk 3 above)
Comment (JCanaks) [reply src/views/articles/subcategory/SingleArticle.js]: Hmm ok. I believe the same component can be used for both viewing a single article and an article in a subcategory as well

Comment (JCanaks) [reply src/views/articles/subcategory/SingleArticle.js]: done

File: src/views/user/Profile.js [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
+
+const Profile = () => (
```
[the remark this answers is not in the document]
Comment (JCanaks) [reply src/views/user/Profile.js:3]: Outdated hound comment. All checks have passed (View PR checks to confirm)

File: src/views/user/Profile.js [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
```
[the remark this answers is not in the document]
Comment (JCanaks) [reply src/views/user/Profile.js:1]: Outdated hound comment. All checks have passed (View PR checks to confirm)

File: src/views/user/Following.js [hunk 7]
```diff
@@ -0,0 +1,7 @@
+import React from 'react';
```
[the remark this answers is not in the document]
Comment (JCanaks) [reply src/views/user/Following.js:1]: Outdated hound comment. All checks have passed (View PR checks to confirm)

File: src/views/auth/AuthRoute.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+const AuthRoute = ({ component: Component, ...rest }) => (
+  <Route
+    {...rest}
+    render={props => (
+      AuthService.isAuthenticated === true
+        ? <Component {...props} />
+        : (
+          <Redirect to={{
+            pathname: '/login',
+            state: { from: props.location }
```
[the remark this answers is not in the document]
Comment (JCanaks) [reply src/views/auth/AuthRoute.js:26]: Outdated hound comment. All checks have passed (View PR checks to confirm)

File: src/views/auth/AuthRoute.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    this.isAuthenticated = true;
+    setTimeout(cb, 100);
+  },
+  logout(cb) {
+    this.isAuthenticated = false;
+    setTimeout(cb, 100);
+  }
+};
+
+const AuthRoute = ({ component: Component, ...rest }) => (
```
[the remark this answers is not in the document]
Comment (JCanaks) [reply src/views/auth/AuthRoute.js:17]: Outdated hound comment. All checks have passed (View PR checks to confirm)

File: src/views/auth/AuthRoute.js (hunk 4 above)
Comment (mppdg) [reply src/views/auth/AuthRoute.js:5]: And as it stands now, it is a dummy representation of the App State. The ticket required that we have a dummy login route setup. After the Redux setup has been merged,  the person assigned the login feature will have to upgrade it accordingly.

File: .eslintrc.json (hunk 1 above)
Comment (Fattylee) [reply .eslintrc.json:9]: OK

File: src/views/auth/AuthRoute.js (hunk 4 above)
Comment (megame24) [reply src/views/auth/AuthRoute.js:5]: alright

**Review by megame24 — APPROVED** [review]

File: src/components/Footer.js [hunk 10]
```diff
@@ -0,0 +1,10 @@
+import React from 'react';
+
+const Footer = () => (
+  <div>
+    <h2>Author's Haven Footer </h2>
+  </div>
+);
+
+
+export default Footer;
```
Comment (tutugodfrey) [inline src/components/Footer.js]: Nice job completing this task, but you need to check your files for EOL's and add them.

Comment (JCanaks) [reply src/components/Footer.js]: No EOL present

File: src/index.js [hunk 11]
```diff
@@ -1,6 +1,9 @@
 import React from 'react';
 import { render } from 'react-dom';
+import { BrowserRouter as Router } from 'react-router-dom';
```
Comment (rhenshaw56) [inline src/index.js:3]: it ain't been used here

Comment (rhenshaw56) [reply src/index.js:3]: take it away

**Review by rhenshaw56 — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 11 above)
Comment (mppdg) [reply src/index.js:3]: That line imports `BrowserRouter` and is given an alias `Router` which is used to wrap the `App` component.
```
<Router>
    <App />
 </Router>
```
Which serves as the parent of all the Routes in the application. Do you mean we should remove the alias and  have something like
```
<BrowserRouter>
    <App />
 </BrowserRouter>
```

Comment (rhenshaw56) [reply src/index.js:3]: right!

**Review by rhenshaw56 — APPROVED** [review]

**Pull request merged**